include "cscbase.thrift"
include "types/cremaemailtypeenum.inc"
include "types/cremauserlevelenum.inc"
include "types/cremacloseenum.inc"

namespace cpp csc.crema
namespace csharp Csc.Crema

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// OpenCremaOrderRequest
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

struct OpenCremaOrderRequest
{
    1:  optional string                 Uuid,
    2:  optional i32                    Country,
    3:  optional i32                    Language,
}

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// OpenCremaOrderResponse
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

struct GuiData
{
    1:  optional i16                    BranchNo,
    2:  optional i32                    Idf,
    3:  optional double                 CreditLimit,
    4:  optional string                 Customer,
    5:  optional i32                    OrderID,
    6:  optional string                 Place,
    7:  optional string                 PostalCode,
    8:  optional string                 Apothecary,
    9:  optional double                 OrderAmount,
    10: optional double                 SumDebitside,
    11: optional double                 SumCreditside,
    12: optional double                 SumMonTurnover,
    13: optional double                 SumDayTurnover,
    14: optional double                 SumReturns,
    15: optional double                 SumGoodsReceipt,
    16: optional i32                    ChainNo,
}

struct MailData
{
    1:  optional string                 Sender,
    2:  optional RecipientList          Recipient_List,
}

struct Recipient
{
    1:  optional cremaemailtypeenum.CremaEmailTypeEnum EmailType,
    2:  optional string                 Address,
}

typedef list<Recipient> RecipientList

struct Reasons
{
    1:  optional list<string>           Reasons_List,
}

struct ApprovalData
{
    1:  optional i32                    ApprovalMax,
    2:  optional i32                    ApprovalClose,
    3:  optional i32                    ApprovalCancel,
}

struct CreditLimitData
{
    1: optional double                 CreditLimit,
    2: optional double                 SumDebitside,
    3: optional double                 SumCreditside,
    4: optional double                 SumMonTurnover,
    5: optional double                 SumDayTurnover,
    6: optional double                 SumReturns,
    7: optional double                 SumGoodsReceipt,
    8: optional i16                    LimitType,
    9: optional bool                   IsChain,
}

typedef list<CreditLimitData> CreditLimitDataList

struct OpenCremaOrderResponse
{
    1:  optional cscbase.ReturnStruct   RetValue,
    2:  optional GuiData                Gui,
    3:  optional MailData               Mail,
    4:  optional Reasons                UnlockReasons,
    5:  optional Reasons                CancelReasons,
    6:  optional ApprovalData           Approval,
    7:  optional CreditLimitDataList    CreditLimits,
}

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// UserSuggestionRequest
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

struct UserSuggestionRequest
{
    1:  optional i16                    BranchNo,
    2:  optional i32                    CustomerNo,
    3:  optional string                 UserName,
    4:  optional bool                   ByTime,
}

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// UserSuggestionResponse
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

struct UserSuggestionResponse
{
    1:  optional cscbase.ReturnStruct   RetValue,
    2:  optional i16                    Choice,
    3:  optional i16                    ChoiceReason,
    4:  optional string                 ChoiceComment,
}

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// CremaOrder
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

struct CremaOrder
{
    1:  optional i32                    OrderNo,
    2:  optional string                 Guid,
    3:  optional cscbase.Date           SendDate,
    4:  optional i16                    BranchNo,
    5:  optional double                 OrderValue,
    6:  optional i32                    CustomerNo,
    7:  optional string                 Url,
    8:  optional string                 ReleaseUser,
    9:  optional string                 ReleaseReason,
    10: optional string                 ReleaseComment,
    11: optional string                 Pharmacy,
    12: optional string                 Place,
    13: optional i32                    ChainNo,
}

typedef list<CremaOrder> CremaOrderList

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// OpenCremaOrdersRequest
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

struct OpenCremaOrdersRequest
{
    1:  optional cremauserlevelenum.CremaUserLevelEnum UserLevel,
    2:  optional list<i16>              RelevantBranchNos,
}

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// OpenCremaOrdersResponse
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

struct OpenCremaOrdersResponse
{
    1:  optional cscbase.ReturnStruct   RetValue,
    2:  optional CremaOrderList         Order_List,
}

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// FinishedTodayCremaOrdersRequest
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

struct FinishedTodayCremaOrdersRequest
{
    1:  optional list<i16>              RelevantBranchNos,
}

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// FinishedTodayCremaOrdersResponse
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

struct FinishedTodayCremaOrdersResponse
{
    1:  optional cscbase.ReturnStruct   RetValue,
    2:  optional CremaOrderList         Order_List,
}

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// LastDecisionRequest
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

struct LastDecisionRequest
{
    1:  optional cremacloseenum.CremaCloseEnum TypeClose,
    2:  optional i32                    Reason,
    3:  optional string                 Comment,
    4:  optional string                 User,
    5:  optional i16                    BranchNo,
    6:  optional i32                    OrderNo,
}

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// LastDecisionResponse
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

struct LastDecisionResponse
{
    1:  optional cscbase.ReturnStruct   RetValue,
}

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// FinishOrderRequest
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

struct FinishOrderRequest
{
    1:  optional i32                    OrderId,
    2:  optional string                 Reason,
    3:  optional string                 Comment,
    4:  optional string                 User,
    5:  optional string                 Guid,
    6:  optional string                 Url,
    7:  optional cremacloseenum.CremaCloseEnum TypeClose,
}

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// FinishOrderResponse
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

struct FinishOrderResponse
{
    1:  optional cscbase.ReturnStruct   RetValue,
}
